google-cloud-spanner/src/main/java/com/google/cloud/spanner/Operation.java

```diff
@@ -120,43 +135,57 @@
    *
    * @return null if operation is not found otherwise the current operation.
    */
-  public Operation<R, M> waitFor(WaitForOption... options) throws SpannerException {
-    if (isDone) {
+  public Operation<R, M> waitFor(RetryOption... waitOptions) throws SpannerException {
+    if (isDone()) {
       return this;
     }
-    long timeoutMillis = Timeout.getOrDefault(options).getTimeoutMillis();
-    boolean hasTimeout = timeoutMillis != -1;
-    CheckingPeriod period = CheckingPeriod.getOrDefault(options);
-    long startMillis = clock.millisTime();
-    while (true) {
-      try {
-        com.google.longrunning.Operation proto = rpc.getOperation(name);
-        if (proto.getDone()) {
-          return Operation.<R, M>create(rpc, proto, parser);
-        }
-        long elapsed = clock.millisTime() - startMillis;
-        if (hasTimeout && elapsed >= timeoutMillis) {
-          throw SpannerExceptionFactory.newSpannerException(
-              ErrorCode.DEADLINE_EXCEEDED, "Operation did not complete in the given time");
-        }
-      } catch (SpannerException e) {
-        if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
-          return null;
-        }
-        if (!e.isRetryable()) {
-          throw e;
-        }
+    RetrySettings waitSettings =
+        RetryOption.mergeToSettings(DEFAULT_OPERATION_WAIT_SETTINGS, waitOptions);
+    try {
+      com.google.longrunning.Operation proto = RetryHelper.poll(
+          new Callable<com.google.longrunning.Operation>() {
+            @Override
+            public com.google.longrunning.Operation call() throws Exception {
+              return rpc.getOperation(name);
+            }
+          },
+          waitSettings,
+          new BasicResultRetryAlgorithm<com.google.longrunning.Operation>() {
+            @Override
+            public boolean shouldRetry(
+                Throwable prevThrowable, com.google.longrunning.Operation prevResponse) {
+              if (prevResponse != null) {
+                return !prevResponse.getDone();
+              }
+              if (prevThrowable instanceof BaseServiceException) {
+                if (!(prevThrowable instanceof SpannerException)) {
+                  return ((BaseServiceException) prevThrowable).isRetryable();
+                }
+                return ((SpannerException) prevThrowable).getErrorCode() != ErrorCode.NOT_FOUND;
```

Comment:
I don't think this is correct. We should only retry if SpannerException.isRetryable is true otherwise we should just bubble up the exception.

Reply:
Seems reasonable, but the old code was doing this:

        if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
          return null;
        }
        if (!e.isRetryable()) {
          throw e;
        }

I.e. it was essentially terminating in case if the error code was not found, and this is what is done here.

Reply:
In the old code, if the exception is SpannerException and is retryable, it would retry. Here if the exception is SpannerException and is not NOT_FOUND it would retry since it returns true in that case.

Reply:
Yes, you are right. I reversed the up the order of condition checks, because I thought it would make code shorter. But it just made it wrong =).